cosmwasm/contracts/wormchain-accounting/src/contract.rs

```diff
@@ -0,0 +1,469 @@
+            ..
+        } => transfer::Data {
+            amount: Uint256::from_be_bytes(amount.0),
+            token_address: TokenAddress::new(token_address.0),
+            token_chain: token_chain.into(),
+            recipient_chain: recipient_chain.into(),
+        },
+        _ => bail!("Unknown tokenbridge payload"),
+    };
+
+    let emitter_chain = o.key.emitter_chain();
+
+    let tokenbridge_addr = TOKENBRIDGE_ADDR
+        .load(deps.storage)
+        .context("failed to load tokenbridge addr")?;
+
+    let registered_emitter: Vec<u8> = deps
+        .querier
+        .query_wasm_smart(
+            tokenbridge_addr,
+            &tokenbridge::msg::QueryMsg::ChainRegistration {
+                chain: emitter_chain,
+            },
+        )
+        .context("failed to query chain registration")?;
+    ensure!(
+        *registered_emitter == **o.key.emitter_address(),
+        "unknown emitter address"
+    );
+
+    accounting::commit_transfer(
+        deps.branch(),
+        Transfer {
+            key: o.key.clone(),
+            data: tx_data,
+        },
+    )
+    .context("failed to commit transfer")?;
+
+    // Now that the transfer has been committed, we don't need to keep it in the pending list.
+    key.remove(deps.storage);
+
+    Ok(Some(
+        Event::new("Transfer")
+            .add_attribute("emitter_chain", o.key.emitter_chain().to_string())
+            .add_attribute("emitter_address", o.key.emitter_address().to_string())
+            .add_attribute("sequence", o.key.sequence().to_string())
+            .add_attribute("nonce", o.nonce.to_string())
+            .add_attribute("tx_hash", o.tx_hash.to_base64())
+            .add_attribute("payload", o.payload.to_base64()),
+    ))
+}
+
+fn modify_balance(
+    deps: DepsMut<WormholeQuery>,
+    info: MessageInfo,
+    modification: Binary,
+    guardian_set_index: u32,
+    signatures: Vec<Signature>,
+) -> Result<Response, AnyError> {
+    deps.querier
+        .query::<Empty>(
+            &WormholeQuery::VerifyQuorum {
+                data: modification.clone(),
+                guardian_set_index,
+                signatures: signatures.into_iter().map(From::from).collect(),
+            }
+            .into(),
+        )
+        .context(ContractError::VerifyQuorum)?;
+
+    let msg: Modification = from_binary(&modification).context("failed to parse `Modification`")?;
+
+    let event = Event::new("Modification")
+        .add_attribute("sequence", msg.sequence.to_string())
+        .add_attribute("chain_id", msg.chain_id.to_string())
+        .add_attribute("token_chain", msg.token_chain.to_string())
+        .add_attribute("token_address", msg.token_address.to_string())
+        .add_attribute("kind", msg.kind.to_string())
+        .add_attribute("amount", hex::encode(msg.amount.to_be_bytes()))
+        .add_attribute("reason", msg.reason.clone());
+
+    accounting::modify_balance(deps, msg).context("failed to modify account balance")?;
```

Comment:
Should the event from this method be propagated to the `Event` down below with `add_event` like in `instantiate`? All of the same info from `modify_balance` is shared with this method's event.

Reply:
Done.  I dropped this event in favor of the one from `modify_balance`.  Also added a test for the event.